File: Python/DFS(longest_path).py
class GNode:
    def __init__(self, id, c="W", d=-1, f=-1, p=None):
        self.id = id  # 노드 ID
        self.color = c  # 노드 색 (W: White, G: Gray)

    def __str__(self):
        return self.id

def find_longest_path(G, start, end):
    curr_path = []  # 현재 경로를 저장하는 리스트
    longest_path = ""  # 가장 긴 경로

    def DFS(G, curr):
        curr.color = "G"  # 현재 노드를 방문 중(Gray)으로 표시
        curr_path.append(curr.id)  # 현재 경로에 노드 추가

        if curr == end and len(curr_path) > 2:
            nonlocal longest_path
            curr_path_str = " ".join(curr_path)
            if len(curr_path_str) > len(longest_path):
                longest_path = curr_path_str  # 가장 긴 경로 업데이트

        for nei in G[curr]:
            if nei.color == "W":
                DFS(G, nei)  # 인접한 미방문 노드를 재귀적으로 탐색

        curr.color = "W"  # 현재 노드를 방문 완료(White)로 표시
        curr_path.pop()  # 현재 경로에서 노드 제거

    DFS(G, start)

    return longest_path if longest_path else "No path found"

r = GNode("r")
s = GNode("s")
t = GNode("t")
u = GNode("u")
v = GNode("v")
w = GNode("w")
x = GNode("x")
y = GNode("y") 

# adj list
G = dict()
G[r], G[w], G[t], G[u] = [w, v], [s, r, t], [s, x, w], [y]
G[v], G[s], G[x], G[y] = [r], [w, t, x], [s, t], [u]


longest_path = find_longest_path(G, t, s)

print("Longest Path:", longest_path)

File: c++/DFS(longest_path).cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <queue>
#include <algorithm>

using namespace std;

class Vertex {
private:
    string name_;
    vector<Vertex*> adjList_;

public:
    Vertex(string name) { name_ = name; }
    string GetName() { return name_; }

    void addToadjList(Vertex* vertex) {
        adjList_.push_back(vertex);
    }
    vector<Vertex*> GetAdjList() { return adjList_; }
};

class Graph {
private:
    vector<Vertex*> vertices_;

public:
    Graph() {}
    ~Graph() {
        for (auto vertex : vertices_)
            delete vertex;
        vertices_.clear();
    }
    Vertex* GenVertex(string name) {
        Vertex* vertex = new Vertex(name);
        vertices_.push_back(vertex);
        return vertex;
    }
    void GenEdge(Vertex* start, Vertex* end) {
        start->addToadjList(end);
        end->addToadjList(start); // Undirected graph, so add the reverse edge
    }
    vector<Vertex*> GetVertices() { return vertices_; }

    // Modified DFS to find the longest path(DFS를 활용해 Path를 저장하는 함수)
    vector<Vertex*> FindLongestPath(Vertex* start, Vertex* end) {
        vector<Vertex*> longestPath;
        vector<Vertex*> currentPath;
        map<Vertex*, bool> visited;

        DFSLongestPath(start, end, currentPath, longestPath, visited);

        return longestPath;
    }

private:
    void DFSLongestPath(Vertex* current, Vertex* end, vector<Vertex*>& currentPath, vector<Vertex*>& longestPath, map<Vertex*, bool>& visited) {
        visited[current] = true;
        currentPath.push_back(current);

        if (current == end && currentPath.size() > longestPath.size()) {
            longestPath = currentPath;
        }

        for (Vertex* neighbor : current->GetAdjList()) {
            if (!visited[neighbor]) {
                DFSLongestPath(neighbor, end, currentPath, longestPath, visited);
            }
        }

        visited[current] = false;   // Backtracking
        currentPath.pop_back();     // Backtracking
    }
};

int main() {
    Graph graph;

    // Generate vertices
    Vertex* vertex_r = graph.GenVertex("r");
    Vertex* vertex_s = graph.GenVertex("s");
    Vertex* vertex_t = graph.GenVertex("t");
    Vertex* vertex_u = graph.GenVertex("u");
    Vertex* vertex_v = graph.GenVertex("v");
    Vertex* vertex_w = graph.GenVertex("w");
    Vertex* vertex_x = graph.GenVertex("x");
    Vertex* vertex_y = graph.GenVertex("y");

    // Generate edges (Undirected Graph)
    graph.GenEdge(vertex_r, vertex_s);
    graph.GenEdge(vertex_s, vertex_r);
    graph.GenEdge(vertex_r, vertex_v);
    graph.GenEdge(vertex_v, vertex_r);
    graph.GenEdge(vertex_s, vertex_w);
    graph.GenEdge(vertex_w, vertex_s);
    graph.GenEdge(vertex_t, vertex_w);
    graph.GenEdge(vertex_w, vertex_t);
    graph.GenEdge(vertex_t, vertex_x);
    graph.GenEdge(vertex_x, vertex_t);
    graph.GenEdge(vertex_t, vertex_u);
    graph.GenEdge(vertex_u, vertex_t);
    graph.GenEdge(vertex_u, vertex_x);
    graph.GenEdge(vertex_x, vertex_u);
    graph.GenEdge(vertex_u, vertex_y);
    graph.GenEdge(vertex_y, vertex_u);
    graph.GenEdge(vertex_w, vertex_x);
    graph.GenEdge(vertex_x, vertex_w);
    graph.GenEdge(vertex_x, vertex_y);
    graph.GenEdge(vertex_y, vertex_x);

    Vertex* start = vertex_s;
    Vertex* end = vertex_y;

    // Find the longest path
    vector<Vertex*> longestPath = graph.FindLongestPath(start, end);

    if (!longestPath.empty()) {
        cout << "Longest Path from " << start->GetName() << " to " << end->GetName() << ": ";
        for (Vertex* vertex : longestPath) {
            cout << vertex->GetName() << " -> ";
        }
        cout << endl;
    } else {
        cout << "No path found." << endl;
    }

    // Clean up memory (delete vertices)
    for (Vertex* vertex : graph.GetVertices()) {
        delete vertex;
    }

    return 0;
}
